### PR TITLE
engine: resources: Change Res API from Compare to Cmp

### PR DESCRIPTION
**Pull request by dbakong**
This will be done by refactoring the current
	the method to return an error message instead
	of a boolean value

## Tips:

* please read the style guide before submitting your patch:
[docs/style-guide.md](../docs/style-guide.md)

* commit message titles must be in the form:

```topic: Capitalized message with no trailing period```

or:

```topic, topic2: Capitalized message with no trailing period```

* golang code must be formatted according to the standard, please run:

```
make gofmt		# formats the entire project correctly
```

or format a single golang file correctly:

```
gofmt -w yourcode.go
```

* please rebase your patch against current git master:

```
git checkout master
git pull origin master
git checkout your-feature
git rebase master
git push your-remote your-feature
hub pull-request	# or submit with the github web ui
```

* after a patch review, please ping @purpleidea so we know to re-review:

```
# make changes based on reviews...
git add -p		# add new changes
git commit --amend	# combine with existing commit
git push your-remote your-feature -f
# now ping @purpleidea in the github PR since it doesn't notify us automatically
```

### CODE REVIEW

**Comment by dbakong** [comment]
#308

File: engine/resources/net.go [hunk 1]
```diff
@@ -506,34 +506,26 @@ func (obj *NetRes) CheckApply(apply bool) (bool, error) {
...
-		return false
+		return fmt.Errorf("the State differs")
 	}
 	if (obj.Addrs == nil) != (res.Addrs == nil) {
-		return false
+		return fmt.Errorf("the Addrs differs")
 	}
 	if err := util.SortedStrSliceCompare(obj.Addrs, res.Addrs); err != nil {
-		return false
+		return fmt.Errorf("one Addrs isn't specified")
```
Comment (purpleidea) [inline engine/resources/net.go]: one necessarily?

File: engine/resources/nspawn.go [hunk 2]
```diff
@@ -261,35 +261,27 @@ func (obj *NspawnRes) CheckApply(apply bool) (bool, error) {
...
+		return fmt.Errorf("the State differs")
 	}
 
 	// TODO: why is res.svc ever nil?
 	if (obj.svc == nil) != (res.svc == nil) { // xor
-		return false
+		return fmt.Errorf("one svc isn't specified")
 	}
 	if obj.svc != nil && res.svc != nil {
-		if !obj.svc.Compare(res.svc) {
```
Comment (purpleidea) [inline engine/resources/nspawn.go]: Why did you change this, we need to `Cmp` on the sub resource.

File: engine/resources/password.go [hunk 3]
```diff
@@ -295,33 +295,25 @@ func (obj *PasswordRes) CheckApply(apply bool) (bool, error) {
...
 
 	if obj.Length != res.Length {
-		return false
+		return fmt.Errorf("the Length differs")
 	}
 	// TODO: we *could* optimize by allowing CheckApply to move from
 	// saved->!saved, by removing the file, but not likely worth it!
 	if obj.Saved != res.Saved {
-		return false
+		return fmt.Errorf("obj.Saved and res.Saved differ")
```
Comment (purpleidea) [inline engine/resources/password.go]: We shouldn't mention `obj.` or `res.` in our error message, the end user doesn't know anything about our implementation.

File: engine/resources/test.go [hunk 4]
```diff
@@ -199,25 +199,17 @@ func (obj *TestRes) CheckApply(apply bool) (bool, error) {
...
-		return false
+		return fmt.Errorf("not a %s", obj.Kind())
 	}
 	//if obj.Name != res.Name {
 	//	return false
 	//}
 
 	if obj.CompareFail || res.CompareFail {
-		return false
+		return fmt.Errorf("res.CompareFail is true")
```
Comment (purpleidea) [inline engine/resources/test.go]: See above comment for all of these.

File: engine/resources/user.go [hunk 5]
```diff
@@ -273,68 +273,60 @@ func (obj *UserRes) CheckApply(apply bool) (bool, error) {
...
+		return fmt.Errorf("not a %s", obj.Kind())
 	}
 
 	if obj.State != res.State {
-		return false
+		return fmt.Errorf("the State differs")
 	}
 	if (obj.UID == nil) != (res.UID == nil) {
-		return false
+		return fmt.Errorf("one UID isn't specified")
```
Comment (purpleidea) [inline engine/resources/user.go]: one?
quiz: What is this check doing? ` 	if (obj.UID == nil) != (res.UID == nil) {` ?

File: engine/resources/user.go [hunk 6]
```diff
@@ -273,68 +273,60 @@ func (obj *UserRes) CheckApply(apply bool) (bool, error) {
...
 	}
 	if obj.UID != nil && res.UID != nil {
 		if *obj.UID != *res.UID {
```
Comment (purpleidea) [inline engine/resources/user.go]: Explain this line of code please.

File: engine/resources/user.go (hunk 5 above)
Comment (dbakong) [reply engine/resources/user.go]: checking if both obj.UID and res.UID exist or are null concurrently.

File: engine/resources/user.go (hunk 6 above)
Comment (dbakong) [reply engine/resources/user.go]: The dereferenced pointer values of obj.UID and res.UID are compared.

File: engine/resources/nspawn.go (hunk 2 above)
Comment (dbakong) [reply engine/resources/nspawn.go]: I updated the condition to include the `Cmp` on the sub resource.

**Comment by dbakong** [comment]
@purpleidea, I made the suggested changes to the code and commit message message. 
Please review. Thanks!

File: engine/resources/msg.go [hunk 7]
```diff
@@ -200,36 +200,28 @@ func (obj *MsgRes) CheckApply(apply bool) (bool, error) {
...
+		return fmt.Errorf("the Priority differs")
 	}
 	if len(obj.Fields) != len(res.Fields) {
-		return false
+		return fmt.Errorf("the length of Fields differs")
 	}
 	for field, value := range obj.Fields {
 		if res.Fields[field] != value {
-			return false
+			return fmt.Errorf("the Fields differs")
```
Comment (purpleidea) [inline engine/resources/msg.go]: grammar: "the fields differ" ?
alternatively you could specify which field name if you wanted to.

File: engine/resources/net.go [hunk 8]
```diff
@@ -506,34 +506,26 @@ func (obj *NetRes) CheckApply(apply bool) (bool, error) {
...
+		return fmt.Errorf("not a %s", obj.Kind())
 	}
 
 	if obj.State != res.State {
-		return false
+		return fmt.Errorf("the State differs")
 	}
 	if (obj.Addrs == nil) != (res.Addrs == nil) {
-		return false
+		return fmt.Errorf("the Addrs differs")
```
Comment (purpleidea) [inline engine/resources/net.go]: grammar

File: engine/resources/net.go [hunk 9]
```diff
@@ -506,34 +506,26 @@ func (obj *NetRes) CheckApply(apply bool) (bool, error) {
...
 	}
 	if err := util.SortedStrSliceCompare(obj.Addrs, res.Addrs); err != nil {
-		return false
+		return fmt.Errorf("the Addrs differs")
```
Comment (purpleidea) [inline engine/resources/net.go]: grammar

File: engine/resources/nspawn.go [hunk 10]
```diff
@@ -261,35 +261,27 @@ func (obj *NspawnRes) CheckApply(apply bool) (bool, error) {
...
 	// TODO: why is res.svc ever nil?
 	if (obj.svc == nil) != (res.svc == nil) { // xor
-		return false
+		return fmt.Errorf("the svc differs")
 	}
 	if obj.svc != nil && res.svc != nil {
-		if !obj.svc.Compare(res.svc) {
-			return false
+		if err := obj.svc.Cmp(res.svc); err != nil {
+			return fmt.Errorf("the svc differs")
```
Comment (purpleidea) [inline engine/resources/nspawn.go]: errwrap.Wrapf instead

**Comment by purpleidea** [comment]
I rebased to master and re-flowed the commit message body. For some reason there's weird whitespace indentation there. Shows up in `git log`. GH seems to miss this. Have a look for next time.
Merged, thanks!
